src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1527,6 +1528,54 @@ def reoffset(o):
 
         minimize_int(offset, reoffset)
 
+    def shrink_offset_pairs(self):
+        """Lower any two blocks offset from each other the same ammount.
+
+        Try lowering every non-zero pair of non-shrinking blocks by a
+        common offset.
+        """
+        self.debug('Shrinking offset pairs.')
+
+        def int_from_block(i):
+            u, v = self.blocks[i]
+            return int_from_bytes(self.shrink_target.buffer[u:v])
+
+        def block_len(i):
+            u, v = self.blocks[i]
+            return v - u
+
+        # Try reoffseting every pair
+        def reoffset_pair(pair, o):
+            n = len(self.blocks)
+            # Number of blocks may have changed, need to validate
+            valid_pair = [p for p in pair if p < n and int_from_block(p) > 0]
+
+            if len(valid_pair) < 2:
+                return
+            m = min([int_from_block(p) for p in valid_pair])
+
+            new_blocks = [self.shrink_target.buffer[u:v]
+                          for u, v in self.blocks]
+            for i in valid_pair:
+                new_blocks[i] = int_to_bytes(
+                    int_from_block(i) + o - m, block_len(i))
```

Comment:
Sorry, starting at this code for a bit I noticed another weird gotcha of the shrink process (I'm sorry you keep running into these. IOU some increased documentation of weird things that can happen during shrink!)

One of the things that can happen is that we will sometimes rewrite the byte stream to get a better result for shrinking. Unfortunately what this means here is that the offset can change out from under you: Say you previously had `x == y + 2` but you suddenly found a value of `x` which rewrites `y` to `1`. Your offset has now shrunk from `x` to `1`, and you'll be trying a larger buffer rather than a smaller one.

Here's a test that might see this happening:

```python
@run_to_buffer
def x(data):
    m = data.draw_bits(8)
    if m < 5:
        data.mark_invalid()
    if m == 5:
        data.write(hbytes([1]))
        data.mark_interesting()
    n = data.draw_bits(8)
    if m == n + 2:
        data.mark_interesting()
assert x == hbytes([5])
```

In general the philosophy of writing shrink passes (which I definitely haven't sufficiently explained, sorry!) is that `self.shrink_target` should be treated as able to change in essentially arbitrary ways after the first point in it that you've changed, so some of the shrink passes will save the current `shrink_target` before doing anything. e.g. take a look at [`lower_common_block_offset`](https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/internal/conjecture/engine.py#L1470)

Reply:
Hmm, I wrote a similar (ok, exact) test like this just now, and it seems to be reducing to `[5,1]` instead of `[5]`. It covers the case we were talking about where `self.shrink_target` changes in the middle, but I'm not sure if it should be shrinking to `[5]`. 

I'll be updating the PR shortly if all local testing goes well, maybe you can take a peek and make sure it makes sense.

Reply:
Sorry, that was my bad! `[5, 1]` is the correct reduction.

I'm a little puzzled that the test passes though. I'll need to stare at the code a bit longer. I'd *expect* [this assertion](https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/internal/conjecture/engine.py#L1221) to be firing, but clearly it's not and I'm not sure if I'm just misunderstanding the code or if there's a more subtle bug lurking.

Reply:
In case this helps either of you:

I think it's the `p < n` in is this line that saves Kyle from the assertion (at least in this example):

```
valid_pair = [p for p in pair if p < n and int_from_block(p) > 0]
```

It's when we call `self.incorporate_new_buffer(b'\x05\x03')` that `self.shrink_target.buffer` becomes `b'\x05'` (this mutation happens in `self.__engine.test_function(data)`). 

So then there's only 1 block, and we only try to incorporate one shrunken block from the pair (since `len(valid_pair) == 1` now). Without filtering `valid_pair` like that, I think we would indeed get the `AssertionError` @DRMacIver is talking about (well, if we got past the IndexError we'd get from `int_to_bytes(int_from_block(i) + o - m, 1)`).

I haven't worked out whether there's a similar example (it would have to have more than 2 blocks, I think) that really does fail.

@kreeve: I'm not really competent to comment on this, but I'm confused about `valid_pair`: if it only has one index, then does it make more sense to just give up (`return False`?) than try to reduce the single block without a partner?

Note to anyone playing around with this example: It doesn't always hit Kyle's new code at all. I've been running it until it does and just looking at what's going on in those cases.

Reply:
It looks to me like the 2 things that save Kyle from that assertion are:

- `valid_pair = ...` (this avoid trying anything with more blocks than the current shrink_target) (@kreeve: if only one index from the pair is valid, why try to shrink the single block without its partner? does it make more sense to give up there?)
- `new_blocks = [self.shrink_target.buffer[u:v] ...]` (this means Kyle's always shrinking relative to the current best)

Reply:
The reason it's not hitting that `assert` anymore is because I'm copying the contents of our buffer into `current` before each call into `minimize_int`, and saving the state before any calls to `incorporate_buffer`. 

Regarding `valid_pair`, I'm going to take that out. Given the checks I have in place now, there'll never be fewer than 2 in that list.

Reply:
Scratch that last part, just verified that was happening sometimes. I've changed it to give up if only half the original pair is still valid.